### ISSUE TITLE
getting error and warning messages when running rootless podman via mpirun on multiple hosts

### ISSUE THREAD

**Issue by firefly-serenity** [containers/libpod#5626]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When starting rootless containers via mpirun on multiple hosts, the processes finish with the following messages (a lot of them):

time="2020-03-26T16:47:52+01:00" level=warning msg="no such directory for freezer.state"
ERRO[1198] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"

Am I doing something wrong or is this a bug?

**Steps to reproduce the issue:**

Not that easy because I am testing with Inifiband, multiple hosts, and a lsdyna

1. Calling podman like so:
```
mpirun \
   -IBV -prot \
   -hostfile hostfile \
   podman run \
      --rm \
      --env-host \
           --userns=keep-id \
           --net=host --pid=host --ipc=host \
           --volume /chome/poduser:/chome/poduser \
           --volume /dev/infiniband:/dev/infiniband \
                  localhost/ls-dyna dyna \
                   i=Caravan-V03c-2400k-main-shell16-120ms.k \
                   memory=600m \
                   memory2=60m \
                   endtime=0.02
```

2. Wait until the job is done

**Describe the results you received:**
Multiple lines like this:

time="2020-03-26T16:47:52+01:00" level=warning msg="no such directory for freezer.state"
ERRO[1198] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"

**Describe the results you expected:**

no errors and no warnings

**Additional information you deem important (e.g. issue happens only occasionally):**

I am getting a lot more of the warning lines (one per container) than the error lines.

**Output of `podman version`:**

```
Version:            1.6.4
RemoteAPI Version:  1
Go Version:         go1.12.12
OS/Arch:            linux/amd64
```

**Output of `podman info --debug`:**

```
debug:
  compiler: gc
  git commit: ""
  go version: go1.12.12
  podman version: 1.6.4
host:
  BuildahVersion: 1.12.0-dev
  CgroupVersion: v1
  Conmon:
    package: conmon-2.0.6-1.module+el8.1.1+5259+bcdd613a.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.6, commit: 6ffbb2ec70dbe5ba56e4bfde946fb04f19dd8bbf'
  Distribution:
    distribution: '"centos"'
    version: "8"
  IDMappings:
    gidmap:
    - container_id: 0
      host_id: 1002
      size: 1
    - container_id: 1
      host_id: 165536
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1002
      size: 1
    - container_id: 1
      host_id: 165536
      size: 65536
  MemFree: 77095333888
  MemTotal: 134818603008
  OCIRuntime:
    name: runc
    package: runc-1.0.0-60.rc8.module_el8.1.0+237+63e26edc.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.1-dev'
  SwapFree: 4287610880
  SwapTotal: 4294963200
  arch: amd64
  cpus: 48
  eventlogger: journald
  hostname: smtcac1903
  kernel: 4.18.0-147.3.1.el8_1.x86_64
  os: linux
  rootless: true
  slirp4netns:
    Executable: /bin/slirp4netns
    Package: slirp4netns-0.4.2-2.git21fdece.module+el8.1.1+5460+3ac089c3.x86_64
    Version: |-
      slirp4netns version 0.4.2+dev
      commit: 21fdece2737dc24ffa3f01a341b8a6854f8b13b4
  uptime: 1493h 26m 33.17s (Approximately 62.21 days)
registries:
  blocked: null
  insecure: null
  search: null
store:
  ConfigFile: /chome/poduser/.config/containers/storage.conf
  ContainerStore:
    number: 81
  GraphDriverName: overlay
  GraphOptions:
    overlay.mount_program:
      Executable: /bin/fuse-overlayfs
      Package: fuse-overlayfs-0.4.1-1.module_el8.1.0+237+63e26edc.x86_64
      Version: |-
        fuse-overlayfs: version 0.4.1
        FUSE library version 3.2.1
        using FUSE kernel interface version 7.26
  GraphRoot: /chome/poduser/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 1
  RunRoot: /tmp/run-1002
  VolumePath: /chome/poduser/.local/share/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.6.4-2.module+el8.1.1+5363+bf8ff1af.x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical

**Comment by rhatdan** [comment]
@giuseppe PTAL
This looks like it is trying to use cgroups on a rootless environment

**Comment by mheon** [comment]
`  CgroupVersion: v1`

Info says we aren't in unified mode, so we shouldn't be.

The closed namespace makes me very concerned, though - something might be cleaning up/removing our containers before we get a chance to, which might impact lifecycle.

**Comment by firefly-serenity** [comment]
To add some more info about it. There are exactly 24 ERRO messages, which matches the number of processes (aka containers) started on each node. All the messages combined:

```
ERRO[1194] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1195] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1197] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1197] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1197] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1198] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1198] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1200] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1200] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1200] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1200] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1201] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
ERRO[1208] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
```

**Comment by giuseppe** [comment]
> time="2020-03-26T16:47:52+01:00" level=warning msg="no such directory for freezer.state"

the error is raised by runc.  I could reproduce locally and it doesn't happen with crun.

> ERRO[1198] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"

This one is from Podman but I am unable to reproduce, do you get the same error if run podman manually without using `mpirun`?

**Comment by firefly-serenity** [comment]
> the error is raised by runc. I could reproduce locally and it doesn't happen with crun.
I will install crun on the systems to confirm. Can you tell me why it is, that there is this difference? 

> This one is from Podman but I am unable to reproduce, do you get the same error if run podman manually without using mpirun?
The error does not occure when starting a single container locally, that I can confirm.

**Comment by giuseppe** [comment]
> > the error is raised by runc. I could reproduce locally and it doesn't happen with crun.
> 
> I will install crun on the systems to confirm. Can you tell me why it is, that there is this difference?

when you specify `--pid=host` the OCI runtime tries to use the freezer controller to freeze the container before killing its processes.
On rootless cgroup v1 that is not possible since rootless cannot use cgroups.  You can ignore the warning, or not use `--pid=host`

**Comment by adrianreber** [comment]
One thing which is special when using MPI is that it might not create sessions as a normal login. Not sure if `XDG_RUNTIME_DIR` is set when using `mpirun`.

@firefly-serenity Do you know how your `mpirun` creates the processes on the other hosts? Is it using `ssh`? In case you are running under the control of a resource manager (slurm, pbs, something) it might not use `ssh` but some PMI(X) based mechanism. When I was trying Podman in combination with slurm I made sure to create the following directories using `slurm.prolog`
```
mkdir /run/user/$SLURM_JOB_UID
chown $SLURM_JOB_UID:`id -g $SLURM_JOB_UID` /run/user/$SLURM_JOB_UID
chmod 711 /run/user/$SLURM_JOB_UID
```

**Comment by firefly-serenity** [comment]
An update: 
> the error is raised by runc. I could reproduce locally and it doesn't happen with crun.

I could reporduce this. And it is also slightly faster.

@adrianreber these tests are started manually and mpirun is using ssh.

These messages are now appearing after each job:
```
time="2020-04-05T17:13:07+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:07+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:08+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:09+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:09+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:10+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:11+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:11+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:11+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:13+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:13+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:14+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:15+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:15+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:15+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:15+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:15+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""
time="2020-04-05T17:13:16+02:00" level=error msg="unable to close namespace: \"close /proc/39696/ns/user: bad file descriptor\""

```

**Comment by firefly-serenity** [comment]
I am currently doing some standard benchmarks in within a podman container and came by the same error message. This time there was no mpi involved:

```
poduser@hostname ~ $ for n in 1 2 3 4 5; do podman run --runtime=/usr/local/bin/runc --rm localhost/bench sysbench cpu --threads=24 --cpu-max-prime=40000000 run | grep "total time:"
; done
Error: error attaching to container 1f6607ab0cb498c22561b4854cccbe28960c37efaa81d8b9779047bf876780d6: write /dev/stdout: broken pipe
    total time:                          138.8291s
ERRO[0140] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
    total time:                          137.2535s
ERRO[0138] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
    total time:                          137.2052s
ERRO[0138] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
    total time:                          137.1882s
ERRO[0138] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"
    total time:                          137.2370s
ERRO[0138] unable to close namespace: "close /proc/39696/ns/user: bad file descriptor"

```

I suspect that the error occures when the system needs to deal with a huge load. Because, when doing the same benchmark with a smaller problem size I get no error:

```
poduser@smtcac1903 ~ $ for n in 1 2 3 4 5; do podman run --runtime=/usr/local/bin/runc --rm localhost/bench sysbench cpu --threads=24 --cpu-max-prime=400000 run | grep "total time:" ; done
    total time:                          10.1979s
    total time:                          10.1729s
    total time:                          10.1894s
    total time:                          10.1975s
    total time:                          10.1931s
```

Maybe you can reporduce this.

**Comment by giuseppe** [comment]
what podman version are you using?

The issue you are seeing should be fixed both on master and on the v1.9 branch

**Comment by firefly-serenity** [comment]
Since I am testing for a productive environment I am stuck with the current release that comes with RedHat 8 which is 1.6 I think.

But it is good to know, that the issues have been fixed in later releases. Just out of curiousity, what was the problem that led to the "unable to close namespace" error?

**Comment by giuseppe** [comment]
the patch that fixes it is 788fdc685b00dee5ccb594bef845204250c4c123

You can safely ignore the error message.  What happens is that Podman closes all the files once it re-execs itself in the child user namespace, and the userns/mountns file descriptors were mistakenly closed twice.

**Comment by firefly-serenity** [comment]
Thank you @giuseppe
